tests/test_cli.py

```diff
@@ -517,3 +536,55 @@ def test_run_non_universal_newline(self):
             '\n' % path)
         self.assertEqual(
             (ctx.returncode, ctx.stdout, ctx.stderr), (1, expected_out, ''))
+
+    def test_encoding_detection_utf16le(self):
+        path = os.path.join(self.wd, 'non-ascii/utf-16-le')
+        with RunContext(self) as ctx:
+            cli.run(('-f', 'parsable', path))
+        self.assertEqual((ctx.returncode, ctx.stdout, ctx.stderr), (0, '', ''))
+
+    def test_encoding_detection_utf16be(self):
+        path = os.path.join(self.wd, 'non-ascii/utf-16-be')
+        with RunContext(self) as ctx:
+            cli.run(('-f', 'parsable', path))
+        self.assertEqual((ctx.returncode, ctx.stdout, ctx.stderr), (0, '', ''))
+
+    def test_encoding_detection_utf8(self):
+        path = os.path.join(self.wd, 'non-ascii/utf-8')
+        with RunContext(self) as ctx:
+            cli.run(('-f', 'parsable', path))
+        self.assertEqual((ctx.returncode, ctx.stdout, ctx.stderr), (0, '', ''))
+
+    def test_detected_encoding_utf8(self):
+        path = os.path.join(self.wd, 'non-ascii/éçäγλνπ¥/utf-8')
+        with cli.yamlopen(path) as yaml_file:
+            yaml_file.read()
+        self.assertEqual(yaml_file.encoding, 'utf-8')
+
+    def test_detected_encoding_utf8_sig(self):
+        path = os.path.join(self.wd, 'non-ascii/utf-8')
+        with cli.yamlopen(path) as yaml_file:
+            yaml_file.read()
+        self.assertEqual(yaml_file.encoding, 'UTF-8-SIG')
+
+    def test_detected_encoding_utf16(self):
+        path = os.path.join(self.wd, 'non-ascii/utf-16-le')
+        with cli.yamlopen(path) as yaml_file:
+            yaml_file.read()
+        self.assertEqual(yaml_file.encoding, 'UTF-16')
+        path = os.path.join(self.wd, 'non-ascii/utf-16-be')
+        with cli.yamlopen(path) as yaml_file:
+            yaml_file.read()
+        self.assertEqual(yaml_file.encoding, 'UTF-16')
+
+    def test_explicit_encoding(self):
+        path = os.path.join(self.wd, 'a.yaml')
+        with cli.yamlopen(path, encoding='windows-1252') as yaml_file:
+            yaml_file.read()
+        self.assertEqual(yaml_file.encoding, 'windows-1252')
```

Comment:
This does not test anything because it was already opened with `encoding='windows-1252'`, does it?

Reply:
It tests that `yamlopen` correctly accepts an explicit encoding, instead of detecting the encoding with chardet.

There are [tests that run the linter](https://github.com/adrienverge/yamllint/pull/240/files#diff-09f983a1db5a7aca6102f2e2a260c88cR540-R556) on all the new non-ascii files, except for the file that is made up of random bytes that cannot be decoded and exists just to be sure UTF-8 is used as a default if chardet can't detect any encoding.

I understand not wanting to add dependencies. If it's any consolation chardet itself is pure python with no external dependencies. I can also go back and revert to just checking the BOM, per the yaml spec and expecting failure or malformed data in any other case.

The motivation for using chardet was [the comment from bz2](https://github.com/adrienverge/yamllint/pull/240#issuecomment-608387745) about making a best-effort to parse files even if they're not an acceptable encoding.

The yamlopen context manager was also implemented [per request](https://github.com/adrienverge/yamllint/pull/240#discussion_r397795662) on review. I can remove that, too, if you prefer not to use it.

Reply:
> It tests that `yamlopen` correctly accepts an explicit encoding, instead of detecting the encoding with chardet.

OK, thanks for the explanation!

> There are [tests that run the linter](https://github.com/adrienverge/yamllint/pull/240/files#diff-09f983a1db5a7aca6102f2e2a260c88cR540-R556) on all the new non-ascii files, except for the file that is made up of random bytes that cannot be decoded and exists just to be sure UTF-8 is used as a default if chardet can't detect any encoding.

Sure. But still no tests for ISO-8859-1-encoded file, unless I missed something?

> I understand not wanting to add dependencies. If it's any consolation chardet itself is pure python with no external dependencies. I can also go back and revert to just checking the BOM, per the yaml spec and expecting failure or malformed data in any other case.
> 
> The motivation for using chardet was [the comment from bz2](https://github.com/adrienverge/yamllint/pull/240#issuecomment-608387745) about making a best-effort to parse files even if they're not an acceptable encoding.

Sure, I understood that. But I'm very against adding new dependencies (pure Python or not).
Have you seen my proposal about a temporary Python-2-only solution? What do you think of it?

> The yamlopen context manager was also implemented [per request](https://github.com/adrienverge/yamllint/pull/240#discussion_r397795662) on review. I can remove that, too, if you prefer not to use it.

Yes I've seen it, but I don't think it helps readability.